toggl/mobileapp
# Use DTOs for creating and updating data instead of Builders and Foundation models

Closes #1711 
Closes #1681 

This is a scary PR. Let's acknowledge that first.

Good. Now what this PR does: it separates the data we use for creating and updating entities in our database (either created in the UI or fetched from te server) from the Foundation and `IDatabase` models.

## Why should we do this?

1. To get rid of our old builders, which turned out to be error prone.
2. The foundation models we construct with the builders miss the important relationships to the other issues, they have just the IDs and this is potentially dangerous - somebody might add code which will use the fake `IThreadSafe` model and it will break stuff (null refs, yaaay).

## What does this PR do?

1. For every entity a new DTO struct is introduced, e.g., `struct ClientDto : IClient, IDatabaseSyncable` with several factory methods
     - `Clean`, `Dirty`, `DirtyDeleted`, `Unsyncable` - which we use during sync (both pull and push)
     - `From` for creating the DTO from an existing entity with potentially new values for some properties (using the `New<T>` type)
2. The repositories and data sources are updated so that they take the DTOs as inputs for all of the methods - for `Create`, `Update`, `BatchUpdate` (and also ofc the methods using `BatchUpdate`: `Overwrite`, `OverwriteIfLocalEntityDidNotChange` (or how it is called...), `UpdateWithConflictResolution`, `DeleteAll`) + updating all the generics which are necessary (that is probably the most confusing part of this PR)
3. Updating conflict and rivals resolution because it now has to play nicely with these DTOs
4. Removing all of the unnecessary code from the Foundation model classes and constructors
5. Using the DTOs all over Foundation instead of the builders
6. Making tests work (in many cases using mock models instead of the builders to create threadsafe model instances)

## What this PR does NOT do?

1. Introduce interactors to remove code from data sources
2. Remove DTOs which are passed to the data sources
3. Remove dependency between the threadsafe and database models or change any of the older model types

## How to review this?

A good question.

- First, I would recommend to have a look at the new DTO structs in the first commit: [0fd4c40](https://github.com/toggl/mobileapp/commit/0fd4c40b54b9ef766b34ba59450d27335e0512ad)
- Then go commit by commit and look what was changed
    - check the change of the signatures of the repositories and how the generics are changed - could this be somehow simplified? I couldn't think of any way...
    - check the same for data sources
- When you get to the ✅ tests, be patient and see that most of the changes are trivial
    - please make sure I did not mess up the tests for conflict resolution
- Think if there are more things we could 🔥 without breaking anything - removing code is fun! Let's remove as much as possible!

## Is this everything?

There should be some follow-ups to this refactoring later:

1. Remove the ID properties from both the `IThreadSafe*` and `IDatabase*`  models - we should get the IDs by accessing the entities, not these properties (i.e., the models maybe should not really implement the Multivac interfaces, which is a bit controversial I think).

2. The `IThreadSafe*` models should not have references to the database models... it just doesn't make sesnse and the only reason they are there is because it implements the `IDatabase*` interface - there should be a conversion instead of this inheritance.

## Are you still scared of this PR?

I will gladly schedule a call with you and explain this PR in more detail or answer any of your questions.

## Can we merge this ASAP?

Probably not - we should wait for the Android public release before making such "drastic" changes. Also I will be on vacation most of the July, so I will probably be able to address your reviews when I come back in August (I expect many requested changes...).
